### PR TITLE
LBFGS optimizer

### PR DESCRIPTION
**Pull request by marpulli**
*Issue #, if available:* #75

*Description of changes:* This is a draft pull request to add LBFGS optimizer for optimization of deterministic loss functions. I found this works much better than the current default optimizer for vanilla GPs.

If you don't object to the approach I use here I will take the time and add tests and tidy the code up.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: mxfusion/inference/batch_loop.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 # ==============================================================================
 
 import mxnet as mx
+from scipy.optimize import fmin_l_bfgs_b
```
Comment (meissnereric) [inline mxfusion/inference/batch_loop.py]: We won't want to take a hard requirement on scipy, so if you can push this import into the functions that use this function that is best. Looks like just the "run" function.

File: mxfusion/inference/batch_loop.py [hunk 2]
```diff
@@ -66,3 +67,97 @@ def run(self, infr_executor, data, param_dict, ctx, optimizer='adam',
...
+
+            i += val.data().size
+            self._shapes.append(val.data().shape)
+
+    def _set_param_values_from_array(self, x, param_dict):
+        """
+        Sets the updated x values in the parameter dictionary
+        """
+        for i, (key, val) in enumerate(param_dict.items()):
+            value_to_set_to = mx.nd.reshape(mx.nd.array(x[self._i_start[i]:self._i_end[i]]), self._shapes[i])
```
Comment (zhenwendai) [inline mxfusion/inference/batch_loop.py]: be careful about the dtype and ctx of the mxnet array.

File: mxfusion/inference/batch_loop.py [hunk 3]
```diff
@@ -66,3 +67,97 @@ def run(self, infr_executor, data, param_dict, ctx, optimizer='adam',
...
+            val.set_data(value_to_set_to)
+
+    def _extract_array_from_param_dict(self, param_dict):
+        """
+        Extracts the flattened parameter array from the parameter dictionary
+        """
+        x = mx.nd.zeros(self._i_end[-1])
+        for i, (key, val) in enumerate(param_dict.items()):
+            x[self._i_start[i]:self._i_end[i]] = val.data().reshape((-1,))
```
Comment (zhenwendai) [inline mxfusion/inference/batch_loop.py]: What about fixed variable?

File: mxfusion/inference/batch_loop.py [hunk 4]
```diff
@@ -66,3 +67,97 @@ def run(self, infr_executor, data, param_dict, ctx, optimizer='adam',
...
+        x = mx.nd.zeros(self._i_end[-1])
+        for i, (key, val) in enumerate(param_dict.items()):
+            x[self._i_start[i]:self._i_end[i]] = val.data().grad.reshape((-1,))
+        return x.asnumpy()
+
+    def _initialize(self, param_dict):
+        """
+        Stores stuff like shapes of all the parameters so we can flatten them all to one array for scipy and reshape
+        them back to the correct shape
+        """
```
Comment (zhenwendai) [inline mxfusion/inference/batch_loop.py]: Initialize and cache the shape and offset of individual parameters seem unnecessary and prone to error, e.g., a user may fix or unfix an array between runs. I would suggest to compute this on-the-fly when the run function is called.

File: mxfusion/inference/batch_loop.py [hunk 5]
```diff
@@ -66,3 +67,97 @@ def run(self, infr_executor, data, param_dict, ctx, optimizer='adam',
...
+        :type verbose: boolean
+        :param logger: The logger to send logs to
+        :type logger: :class:`inference.Logger`
+        """
+
+        if self._initialized is None:
+            self._initialize(param_dict)
+
+        x0 = self._extract_array_from_param_dict(param_dict)
+        func = lambda x: self._f_df(x, infr_executor, data, ctx, param_dict)
```
Comment (zhenwendai) [inline mxfusion/inference/batch_loop.py]: You can directly create an inner function.

**Review by zhenwendai — COMMENTED** [review]
Thanks for contributing a l-bfgs optimizer.

File: mxfusion/inference/batch_loop.py (hunk 5 above)
Comment (marpulli) [reply mxfusion/inference/batch_loop.py]: Do you mean use `def` instead of a lambda?

File: mxfusion/inference/batch_loop.py (hunk 4 above)
Comment (marpulli) [reply mxfusion/inference/batch_loop.py]: Good point, I'll compute these things every time the `run` method is called.

File: mxfusion/inference/batch_loop.py (hunk 3 above)
Comment (marpulli) [reply mxfusion/inference/batch_loop.py]: Yep, I'll add a check for this.

File: mxfusion/inference/batch_loop.py (hunk 2 above)
Comment (marpulli) [reply mxfusion/inference/batch_loop.py]: Good point. I'll extract the dtype and context of the parameters along with the shapes

File: mxfusion/inference/batch_loop.py (hunk 1 above)
Comment (marpulli) [reply mxfusion/inference/batch_loop.py]: Sure, I'll move this to `run`

File: mxfusion/inference/batch_loop.py (hunk 5 above)
Comment (zhenwendai) [reply mxfusion/inference/batch_loop.py]: Yes. Then, it is easy to put all the content in ```self._f_df``` into the inner function.

**Review by zhenwendai — APPROVED** [review]

**Comment by zhenwendai** [comment]
@marpulli Is this pull request ready for merge?

**Comment by marpulli** [comment]
My only open question is whether `BatchInferenceLoopScipy` is actually a good name as it only let's you use one specific scipy optimizer. I think `BatchInferenceLoopLbfgs` would be better?

**Comment by zhenwendai** [comment]
```BatchInferenceLoopLbfgs``` is better. It should be ```BatchInferenceLoopLBFGS```.

**Comment by marpulli** [comment]
Now renamed, I think it's ready to merge

**Pull request merged**